modules/PhenoDigm.py

```diff
@@ -0,0 +1,362 @@
+            if start == 0:  # Only write the header for the first requested batch.
+                tmp_file.write(header + '\n')
+            number_of_records = 0
+            for line in response_lines:
+                number_of_records += 1
+                tmp_file.write(line + '\n')
+            return number_of_records, tmp_file.name
+
+    def fetch_data(self, data_type, output_filename):
+        """Fetch all rows of the requested data type to the specified location."""
+        total_records = self.get_number_of_solr_records(data_type)
+        assert total_records != 0, f'SOLR did not return any data for {data_type}.'
+        with open(output_filename, 'wb') as outfile:
+            start, total = 0, 0  # Initialise the counters.
+            while True:
+                number_of_records, tmp_filename = self.query_solr(data_type, start)
+                with open(tmp_filename, 'rb') as tmp_file:
+                    shutil.copyfileobj(tmp_file, outfile)
+                os.remove(tmp_filename)
+                # Increment the counters.
+                start += self.rows
+                total += number_of_records
+                # Exit when all documents have been retrieved.
+                if total == total_records:
+                    break
+
+
+class PhenoDigm:
+    """Retrieve the data, load it into Spark, process and write the resulting evidence strings."""
+
+    def __init__(self, logger, cache_dir):
+        super(PhenoDigm, self).__init__()
+        self.logger = logger
+        self.cache_dir = cache_dir
+        self.spark = pyspark.sql.SparkSession.builder.appName('phenodigm_parser').getOrCreate()
+        self.hgnc_gene_id_to_ensembl_human_gene_id, self.mgi_gene_id_to_ensembl_mouse_gene_id = [None] * 2
+        self.mouse_gene_to_human_gene, self.mouse_phenotype_to_human_phenotype = [None] * 2
+        self.mouse_model, self.disease, self.disease_model_summary, self.ontology = [None] * 4
+        self.evidence = None
+
+    def update_cache(self):
+        """Fetch the Ensembl gene ID and SOLR data into the local cache directory."""
+        pathlib.Path(self.cache_dir).mkdir(parents=False, exist_ok=True)
+
+        self.logger.info('Fetching human gene ID mappings from HGNC.')
+        urllib.request.urlretrieve(HGNC_DATASET_URI, os.path.join(self.cache_dir, HGNC_DATASET_FILENAME))
+
+        self.logger.info('Fetching mouse gene ID mappings from MGI.')
+        urllib.request.urlretrieve(MGI_DATASET_URI, os.path.join(self.cache_dir, MGI_DATASET_FILENAME))
+
+        self.logger.info('Fetching PhenoDigm data from IMPC SOLR.')
+        impc_solr_retriever = ImpcSolrRetriever(solr_host=IMPC_SOLR_HOST, timeout=IMPC_SOLR_TIMEOUT,
+                                                rows=IMPC_SOLR_BATCH_SIZE)
+        for data_type in IMPC_SOLR_TABLES:
+            self.logger.info(f'Fetching PhenoDigm data type {data_type}.')
+            filename = os.path.join(self.cache_dir, IMPC_FILENAME.format(data_type=data_type))
```

Comment:
Use f-strings instead.

Reply:
I'm formatting a constant here so it cannot be replaced by an f-string, unless I misunderstand something. Could you provide an example of what you're proposing please?